packages/askar/src/storage/AskarStorageService.ts

```diff
@@ -0,0 +1,291 @@
+      // Otherwise just use the value
+      else {
+        transformedTags[key] = value as string
+      }
+    }
+
+    return transformedTags
+  }
+
+  private transformFromRecordTagValues(tags: TagsBase): { [key: string]: string | undefined } {
+    const transformedTags: { [key: string]: string | undefined } = {}
+
+    for (const [key, value] of Object.entries(tags)) {
+      // If the value is of type null we use the value undefined
+      // Askar doesn't support null as a value
+      if (value === null) {
+        transformedTags[key] = undefined
+      }
+      // If the value is a boolean use the Askar
+      // '1' or '0' syntax
+      else if (typeof value === 'boolean') {
+        transformedTags[key] = value ? '1' : '0'
+      }
+      // If the value is 1 or 0, we need to add something to the value, otherwise
+      // the next time we deserialize the tag values it will be converted to boolean
+      else if (value === '1' || value === '0') {
+        transformedTags[key] = `n__${value}`
+      }
+      // If the value is an array we create a tag for each array
+      // item ("tagName:arrayItem" = "1")
+      else if (Array.isArray(value)) {
+        value.forEach((item) => {
+          const tagName = `${key}:${item}`
+          transformedTags[tagName] = '1'
+        })
+      }
+      // Otherwise just use the value
+      else {
+        transformedTags[key] = value
+      }
+    }
+
+    return transformedTags
+  }
+
+  /**
+   * Transforms the search query into a wallet query compatible with Askar WQL.
+   *
+   * The format used by AFJ is almost the same as the WQL query, with the exception of
+   * the encoding of values, however this is handled by the {@link AskarStorageService.transformToRecordTagValues}
+   * method.
+   */
+  // TODO: Transform to Askar format
+  private askarQueryFromSearchQuery(query: Query<T>): Record<string, unknown> {
+    // eslint-disable-next-line prefer-const
+    let { $and, $or, $not, ...tags } = query
+
+    $and = ($and as Query<T>[] | undefined)?.map((q) => this.askarQueryFromSearchQuery(q))
+    $or = ($or as Query<T>[] | undefined)?.map((q) => this.askarQueryFromSearchQuery(q))
+    $not = $not ? this.askarQueryFromSearchQuery($not as Query<T>) : undefined
+
+    const askarQuery = {
+      ...this.transformFromRecordTagValues(tags as unknown as TagsBase),
+      $and,
+      $or,
+      $not,
+    }
+
+    return askarQuery
+  }
+
+  private recordToInstance(record: EntryObject, recordClass: BaseRecordConstructor<T>): T {
+    const instance = JsonTransformer.deserialize<T>(record.value as string, recordClass)
+    instance.id = record.name
+
+    const tags = record.tags ? this.transformToRecordTagValues(record.tags) : {}
+    instance.replaceTags(tags)
+
+    return instance
+  }
+
+  /** @inheritDoc */
+  public async save(agentContext: AgentContext, record: T) {
+    assertAskarWallet(agentContext.wallet)
+    const session = (agentContext.wallet as AskarWallet).session
```

Comment:
I don't this you need to cast here. The `assertAskarWallet` returns `asserts wallet is AskarWallet` so after calling that method typescript should know the wallet is of type `AskarWallet`

Reply:
Good catch! I've added the assert afterwards and didn't consider that fact.